File: reader-app/src/reader/narration/cleanup.ts
import { FOOTER_BAND, HEADER_BAND } from "../core/config.ts";
import type { DocumentBlockType, DocumentHints } from "../core/types.ts";
import { isPageNumber, type ParagraphGroup } from "./paragraph-builder.ts";

export { isPageNumber };

function normalizeKey(text: string): string {
  return text
    .toLowerCase()
    .replace(/[0-9]+/g, "")
    .replace(/[^\p{L}\s]/gu, "")
    .replace(/\s+/g, " ")
    .trim();
}

function inHeaderBand(group: ParagraphGroup): boolean {
  const top = Math.min(...group.bounds.map((box) => box.y));
  return top < HEADER_BAND;
}

function inFooterBand(group: ParagraphGroup): boolean {
  const bottom = Math.max(...group.bounds.map((box) => box.y + box.height));
  return bottom > 1 - FOOTER_BAND;
}

export function classifyGroup(
  group: ParagraphGroup,
  all: ParagraphGroup[],
  hints: DocumentHints,
): DocumentBlockType {
  const text = group.text.trim();
  if (!text) return "unknown";
  if (isPageNumber(text) && (inHeaderBand(group) || inFooterBand(group))) {
    return "page-number";
  }

  const key = normalizeKey(text);
  if (key && inHeaderBand(group) && hints.headerTexts.includes(key)) {
    return "header";
  }
  if (key && inFooterBand(group) && hints.footerTexts.includes(key)) {
    return "footer";
  }

  const median = medianFont(all);
  const small = group.fontSize < median * 0.78;
  if (small && inHeaderBand(group) && text.length < 80) return "header";
  if (small && inFooterBand(group) && text.length < 80) {
    return isPageNumber(text) ? "page-number" : "footer";
  }

  if (
    small &&
    inFooterBand(group) &&
    /^[0-9]+\s+\S/.test(text) &&
    group.fontSize < median * 0.7
  ) {
    return "footnote";
  }

  if (/^(?:references|bibliography|works cited)$/i.test(text)) {
    return "reference";
  }

  if (/^(?:[-•●▪]|[0-9]{1,2}[.)]|[A-Za-z][.)])\s+\S/.test(text)) {
    return "list-item";
  }

  const largest = Math.max(...all.map((item) => item.fontSize));
  if (group.fontSize >= largest * 0.92 && group.fontSize > median * 1.35) {
    return group.bounds[0] && group.bounds[0].y < 0.28 ? "title" : "heading";
  }
  // A ~1.15-1.2x heading/body ratio is ordinary in printed books (this file's
  // own test book uses 12.5pt over 10.5pt, ~1.19x) - too close to body size
  // for a font-size check alone to be reliable, so it's paired with the
  // short-line, no-terminal-punctuation shape a genuine heading also has.
  if (group.fontSize > median * 1.15 && text.length < 90 && !/[.!?]$/.test(text)) {
    return "heading";
  }
  if (/^fig(?:ure)?\.?\s*\d+/i.test(text) || /^table\s+\d+/i.test(text)) {
    return "caption";
  }
  return "paragraph";
}

const BOILERPLATE_SIGNAL_RE =
  /(?:©|copyright\b|all rights reserved\b|isbn\b|published (?:by|as|in)\b|first published\b|library of congress\b|printed in\b|no part of this (?:book|publication)\b|[a-z]+ (?:books?|press|publishing|publishers?) (?:ltd|inc|llc|group)\b)/i;

const LEGAL_NOTICE_RE =
  /(?:work of fiction|resemblance to (?:actual|real) persons|products? of the author'?s imagination|used fictitiously|without (?:the )?(?:prior )?(?:written )?permission|scanning, uploading|registered trademarks?|cover (?:design|art|illustration) by)/i;

// A run-on narrative sentence - a lowercase clause continuing past its own
// line - is the one shape a copyright/imprint page never has, since every
// line there is its own short publisher, address, or rights notice.
const NARRATIVE_FLOW_RE = /[a-z][,;:]?\s+(?:and|but|so|that|which|who|because)\b/i;

// A copyright/imprint page is built almost entirely from short publisher,
// ISBN, and rights-notice lines rather than prose. Any single title uses its
// own publisher's name and address, so this looks at the *shape* shared by
// nearly all such pages - dense boilerplate signals, no narrative sentences -
// instead of matching one publisher's wording.
export function isBoilerplatePage(groups: ParagraphGroup[]): boolean {
  const candidates = groups.filter((group) => group.text.trim().length > 0);
  if (candidates.length < 3) return false;
  const boilerplateLines = candidates.filter((group) =>
    BOILERPLATE_SIGNAL_RE.test(group.text.trim()) || LEGAL_NOTICE_RE.test(group.text),
  );
  // A legal notice ("names, characters, places, and incidents...") reads as
  // a run-on sentence but is itself imprint boilerplate, not narrative.
  const narrative = candidates.filter(
    (group) => NARRATIVE_FLOW_RE.test(group.text) && !LEGAL_NOTICE_RE.test(group.text),
  );
  return boilerplateLines.length / candidates.length >= 0.4 && narrative.length === 0;
}

function medianFont(groups: ParagraphGroup[]): number {
  const sizes = groups.map((group) => group.fontSize).sort((a, b) => a - b);
  if (sizes.length === 0) return 0.02;
  const mid = Math.floor(sizes.length / 2);
  return sizes[mid] ?? 0.02;
}

export function collectHints(
  groups: ParagraphGroup[],
  hints: DocumentHints,
): DocumentHints {
  const next: DocumentHints = {
    headerTexts: [...hints.headerTexts],
    footerTexts: [...hints.footerTexts],
    lastHeading: hints.lastHeading,
  };
  for (const group of groups) {
    const key = normalizeKey(group.text);
    if (!key || key.length < 4) continue;
    if (inHeaderBand(group) && group.text.length < 80) {
      if (!next.headerTexts.includes(key)) next.headerTexts.push(key);
    }
    if (inFooterBand(group) && group.text.length < 80) {
      if (!next.footerTexts.includes(key)) next.footerTexts.push(key);
    }
  }
  return next;
}

const CITATION_NUMERIC = /\[(?:\d+(?:\s*[,;–-]\s*\d+)*)\]/g;
const CITATION_AUTHOR =
  /\((?:[A-Z][A-Za-z-]+(?:\s+(?:et al\.?|& [A-Z][A-Za-z-]+))?,?\s+\d{4}[a-z]?(?:,\s*p+\.?\s*\d+)?)\)/g;
const SUPER_REF = /(?<=\S)[¹²³⁴⁵⁶⁷⁸⁹⁰]+/g;

export function stripCitations(text: string): string {
  return text
    .replace(CITATION_NUMERIC, "")
    .replace(CITATION_AUTHOR, "")
    .replace(SUPER_REF, "")
    .replace(/\s{2,}/g, " ")
    .replace(/\s+([,.;:!?])/g, "$1")
    .trim();
}

export function ordinalListPrefix(
  text: string,
): { spoken: string; rest: string } | null {
  const match = text.match(/^(?:([0-9]{1,2})[.)]|[-•●▪])\s+(.*)$/);
  if (!match) return null;
  const rest = match[2] ?? "";
  const n = match[1] ? Number(match[1]) : null;
  if (n && n >= 1 && n <= 20) {
    const words = [
      "First",
      "Second",
      "Third",
      "Fourth",
      "Fifth",
      "Sixth",
      "Seventh",
      "Eighth",
      "Ninth",
      "Tenth",
      "Eleventh",
      "Twelfth",
      "Thirteenth",
      "Fourteenth",
      "Fifteenth",
      "Sixteenth",
      "Seventeenth",
      "Eighteenth",
      "Nineteenth",
      "Twentieth",
    ];
    return { spoken: `${words[n - 1]}, ${rest}`, rest };
  }
  return { spoken: rest, rest };
}

File: android-app/core/src/main/kotlin/com/evangeline/reader/narration/Cleanup.kt
package com.evangeline.reader.narration

import com.evangeline.reader.model.DocumentBlockType
import com.evangeline.reader.model.DocumentHints
import com.evangeline.reader.model.FOOTER_BAND
import com.evangeline.reader.model.HEADER_BAND

// A bracketed run of numbers: "[1]", "[1, 2, 3]", "[4-6]".
private val CITATION_NUMERIC = Regex("\\[(?:\\d+(?:\\s*[,;–-]\\s*\\d+)*)]")

// A parenthesised author-year reference: "(Smith, 2020)", "(Smith & Jones, 2019)",
// "(Smith et al., 2020, p. 14)". Deliberately narrow so ordinary parentheses
// such as "(n = 42)" survive - over-stripping would silently drop real prose.
private val CITATION_AUTHOR = Regex(
    "\\((?:[A-Z][A-Za-z-]+(?:\\s+(?:et al\\.?|& [A-Z][A-Za-z-]+))?,?\\s+" +
        "\\d{4}[a-z]?(?:,\\s*p+\\.?\\s*\\d+)?)\\)",
)

// Superscript reference marks attached to a word, as in "result¹".
private val SUPER_REF = Regex("(?<=\\S)[¹²³⁴⁵⁶⁷⁸⁹⁰]+")

private val EXTRA_SPACES = Regex("\\s{2,}")
private val WHITESPACE_RUN = Regex("\\s+")
private val SPACE_BEFORE_PUNCTUATION = Regex("\\s+([,.;:!?])")

/**
 * Removes citation marks so they are not read aloud (FR-006).
 *
 * Spacing is repaired afterwards so removing a citation does not leave a gap
 * before the punctuation that followed it.
 */
fun stripCitations(text: String): String =
    text
        .replace(CITATION_NUMERIC, "")
        .replace(CITATION_AUTHOR, "")
        .replace(SUPER_REF, "")
        .replace(EXTRA_SPACES, " ")
        .replace(SPACE_BEFORE_PUNCTUATION, "$1")
        .trim()

// "12", "Page 7", "iv", "XIV", "3 / 12". Roman numerals and arabic digits, with
// an optional "page" prefix and an optional "of total" suffix.
private val PAGE_NUMBER = Regex(
    "^(?:page\\s+)?(?:[ivxlcdm]+|\\d+)(?:\\s*/\\s*\\d+)?$|^[ivxlcdm]+$|^\\d+\\s*$",
    RegexOption.IGNORE_CASE,
)

// Longer than any real page stamp; past this the text is prose that happens to
// start with a numeral.
private const val MAX_PAGE_STAMP_LENGTH = 18

/** Whether a line is a page stamp, and so must not be spoken (FR-004). */
fun isPageNumber(text: String): Boolean {
    val trimmed = text.trim()
    if (trimmed.isEmpty() || trimmed.length > MAX_PAGE_STAMP_LENGTH) return false
    return PAGE_NUMBER.matches(trimmed)
}

private val ORDINAL_WORDS = listOf(
    "First", "Second", "Third", "Fourth", "Fifth", "Sixth", "Seventh",
    "Eighth", "Ninth", "Tenth", "Eleventh", "Twelfth", "Thirteenth",
    "Fourteenth", "Fifteenth", "Sixteenth", "Seventeenth", "Eighteenth",
    "Nineteenth", "Twentieth",
)

// "1. ", "2) ", "- ", "• " - a marker followed by whitespace. The trailing space
// is required, so "3.NoSpace" is prose, not a list item.
private val LIST_MARKER = Regex("^(?:([0-9]{1,2})[.)]|[-•●▪])\\s+(.*)$")

/**
 * A list item split into what should be spoken and the text after its marker.
 *
 * [spoken] leads with the ordinal as a word ("First, Take notes") so a listener
 * hears position without hearing punctuation. Bullets and numbers above twenty
 * drop the marker entirely rather than reading it out.
 */
data class ListPrefix(
    val spoken: String,
    val rest: String,
)

/** Returns null when the text is not a list item at all. */
fun ordinalListPrefix(text: String): ListPrefix? {
    val match = LIST_MARKER.find(text) ?: return null
    val rest = match.groupValues[2]
    val number = match.groupValues[1].toIntOrNull()
    val word = number?.let { ORDINAL_WORDS.getOrNull(it - 1) }
    return if (word != null) ListPrefix("$word, $rest", rest) else ListPrefix(rest, rest)
}

private val DIGITS = Regex("[0-9]+")
private val NON_LETTER = Regex("[^\\p{L}\\s]")

// Page furniture repeats with the page number changing, so digits are dropped
// before comparison: "Page 12" and "Page 13" must collapse to the same key.
private fun normalizeKey(text: String): String =
    text.lowercase()
        .replace(DIGITS, "")
        .replace(NON_LETTER, "")
        .replace(WHITESPACE_RUN, " ")
        .trim()

private const val MIN_HINT_LENGTH = 4
private const val MAX_FURNITURE_LENGTH = 80

private fun ParagraphGroup.inHeaderBand(): Boolean =
    bounds.minOf { it.y } < HEADER_BAND

private fun ParagraphGroup.inFooterBand(): Boolean =
    bounds.maxOf { it.y + it.height } > 1 - FOOTER_BAND

/**
 * Learns which text repeats as a running header or footer across the document,
 * so it can be suppressed from narration (FR-004).
 *
 * Hints accumulate across pages: a line is only furniture if it recurs, and a
 * caller passes the hints gathered so far back in.
 */
fun collectHints(groups: List<ParagraphGroup>, hints: DocumentHints): DocumentHints {
    val headers = hints.headerTexts.toMutableList()
    val footers = hints.footerTexts.toMutableList()
    for (group in groups) {
        val key = normalizeKey(group.text)
        if (key.length < MIN_HINT_LENGTH) continue
        if (group.text.length >= MAX_FURNITURE_LENGTH) continue
        if (group.inHeaderBand() && key !in headers) headers.add(key)
        if (group.inFooterBand() && key !in footers) footers.add(key)
    }
    return DocumentHints(headerTexts = headers, footerTexts = footers)
}

private val REFERENCE_HEADING = Regex("^(?:references|bibliography|works cited)$", RegexOption.IGNORE_CASE)
private val LIST_ITEM = Regex("^(?:[-•●▪]|[0-9]{1,2}[.)]|[A-Za-z][.)])\\s+\\S")
private val FOOTNOTE_START = Regex("^[0-9]+\\s+\\S")
private val CAPTION = Regex("^fig(?:ure)?\\.?\\s*\\d+|^table\\s+\\d+", RegexOption.IGNORE_CASE)
private val ENDS_SENTENCE = Regex("[.!?]$")

private const val SMALL_FONT_RATIO = 0.78
private const val FOOTNOTE_FONT_RATIO = 0.7
private const val TITLE_FONT_RATIO = 0.92
private const val TITLE_MEDIAN_RATIO = 1.35
private const val HEADING_MEDIAN_RATIO = 1.22
private const val TITLE_MAX_Y = 0.28
private const val HEADING_MAX_LENGTH = 90
private const val DEFAULT_FONT_SIZE = 0.02

private fun medianFont(groups: List<ParagraphGroup>): Double {
    val sizes = groups.map { it.fontSize }.sorted()
    if (sizes.isEmpty()) return DEFAULT_FONT_SIZE
    return sizes[sizes.size / 2]
}

/** Furniture recognised because this exact text repeats across pages. */
private fun repeatedFurniture(
    text: String,
    hints: DocumentHints,
    header: Boolean,
    footer: Boolean,
): DocumentBlockType? {
    val key = normalizeKey(text)
    if (key.isEmpty()) return null
    if (header && key in hints.headerTexts) return DocumentBlockType.HEADER
    if (footer && key in hints.footerTexts) return DocumentBlockType.FOOTER
    return null
}

/** Furniture recognised because it sits at a page edge in smaller type. */
private fun edgeFurniture(
    text: String,
    header: Boolean,
    footer: Boolean,
): DocumentBlockType? {
    if (text.length >= MAX_FURNITURE_LENGTH) return null
    if (header) return DocumentBlockType.HEADER
    if (!footer) return null
    return if (isPageNumber(text)) DocumentBlockType.PAGE_NUMBER else DocumentBlockType.FOOTER
}

/** Page furniture: repeated headers, footers and page stamps. Never spoken. */
private fun furnitureType(
    group: ParagraphGroup,
    text: String,
    hints: DocumentHints,
    isSmall: Boolean,
): DocumentBlockType? {
    val header = group.inHeaderBand()
    val footer = group.inFooterBand()
    if (isPageNumber(text) && (header || footer)) return DocumentBlockType.PAGE_NUMBER
    repeatedFurniture(text, hints, header, footer)?.let { return it }
    return if (isSmall) edgeFurniture(text, header, footer) else null
}

/** A heading or title, judged by font size relative to the rest of the page. */
private fun prominenceType(
    group: ParagraphGroup,
    text: String,
    all: List<ParagraphGroup>,
    median: Double,
): DocumentBlockType? {
    val largest = all.maxOfOrNull { it.fontSize } ?: group.fontSize
    if (group.fontSize >= largest * TITLE_FONT_RATIO && group.fontSize > median * TITLE_MEDIAN_RATIO) {
        val top = group.bounds.firstOrNull()?.y
        return if (top != null && top < TITLE_MAX_Y) DocumentBlockType.TITLE else DocumentBlockType.HEADING
    }
    val standsOut = group.fontSize > median * HEADING_MEDIAN_RATIO &&
        text.length < HEADING_MAX_LENGTH &&
        !ENDS_SENTENCE.containsMatchIn(text)
    return if (standsOut) DocumentBlockType.HEADING else null
}

/**
 * Decides what a paragraph group is, which in turn decides whether it is spoken
 * (FR-004). Furniture is checked first so a running header set in body-sized
 * type is still suppressed.
 */
private fun isFootnote(group: ParagraphGroup, text: String, median: Double): Boolean =
    group.inFooterBand() &&
        FOOTNOTE_START.containsMatchIn(text) &&
        group.fontSize < median * FOOTNOTE_FONT_RATIO

/** Structural kinds recognised from the text itself, independent of type size. */
private fun textualType(text: String): DocumentBlockType? = when {
    REFERENCE_HEADING.matches(text) -> DocumentBlockType.REFERENCE
    LIST_ITEM.containsMatchIn(text) -> DocumentBlockType.LIST_ITEM
    else -> null
}

/**
 * Decides what a paragraph group is, which in turn decides whether it is spoken
 * (FR-004). Furniture is checked first so a running header set in body-sized
 * type is still suppressed.
 */
fun classifyGroup(
    group: ParagraphGroup,
    all: List<ParagraphGroup>,
    hints: DocumentHints,
): DocumentBlockType {
    val text = group.text.trim()
    if (text.isEmpty()) return DocumentBlockType.UNKNOWN

    val median = medianFont(all)
    val isSmall = group.fontSize < median * SMALL_FONT_RATIO
    furnitureType(group, text, hints, isSmall)?.let { return it }
    if (isSmall && isFootnote(group, text, median)) return DocumentBlockType.FOOTNOTE
    textualType(text)?.let { return it }
    prominenceType(group, text, all, median)?.let { return it }
    if (CAPTION.containsMatchIn(text)) return DocumentBlockType.CAPTION
    return DocumentBlockType.PARAGRAPH
}
